R/reduce.R

```diff
@@ -46,3 +46,54 @@ reduce_right <- function(.x, .f, ..., .init) {
 
   Reduce(f, .x, init = .init, right = TRUE)
 }
+
+#' Accumulate recursive folds across a list
+#'
+#' \code{accumulate} applies a function recursively over a list from the left, while
+#' \code{accumulate_right} applies the function from the right. Unlike \code{reduce}
+#' both functions keep the intermediate results.
+#'
+#' @inheritParams reduce
+#' @export
+#'
+#' @examples
+#' 1:3 %>% accumulate(`+`)
+#' 1:10 %>% accumulate_right(`*`)
+#'
+#' # From Haskell's scanl documentation
+#' 1:10 %>% accumulate(max, .init = 5)
+#'
+#' # Can be used to floor and cap
+#' caps <- mtcars %>% map(~ quantile(.x, probs = c(.1, .9), names = FALSE))
+#' floor_cap <- mtcars %>% map2(caps, ~ accumulate(.x, max, .init = .y[1])) %>%
+#'   map2(caps, ~ accumulate(.x, min, .init = .y[2]))
+#'
+#' # Simulating stochastic processes with drift
+#' \dontrun{
+#' library(dplyr)
+#' library(ggplot2)
+#'
+#' rerun(5, rnorm(100)) %>% setNames(paste0("sim", 1:5)) %>%
+#'   map(~ accumulate(., function(.x, .y) .05 + .x + .y)) %>%
+#'   map(~ data.frame(value = .x, step = 1:100)) %>%
```

Comment:
This could now be `map_df(~ data_frame(value = .x, step = 1:100), .id = "simulation")`